File: Service.Models/Bikes/BikeListItem.cs
using System;
using System.ComponentModel.DataAnnotations;
using Toptal.BikeRentals.BusinessEntities.Bikes;
using Toptal.BikeRentals.BusinessEntities.Master;

namespace Toptal.BikeRentals.Service.Models.Bikes
{
    public class BikeListItem : BikeRef
    {
        [Required]
        public BikeState? BikeState { get; set; }

        [Required]
        public new BikeModel BikeModel { get { return (BikeModel)base.BikeModel; } set { base.BikeModel = value; } }

        [Required]
        public Location CurrentLocation { get; set; }

        [Required]
        [StringLength(100)]
        public string CurrentLocationName { get; set; }
        
        public DateTime? AvailableFromUtc { get; set; }

        public double? DistanceMiles { get; private set; }

        public float? RateAverage { get; set; }

        public BikeListItem(BusinessEntities.Bikes.Bike bike, Location? currentLocation)
            : base(bike)
        {
            if (bike != null)
            {
                this.BikeModel = new BikeModel(bike.BikeModel);
                this.BikeState = bike.BikeState;
                this.CurrentLocation = bike.CurrentLocation;
                this.AvailableFromUtc = bike.AvailableFromUtc;
                this.RateAverage = bike.RateAverage;
                this.CurrentLocationName = bike.CurrentLocationName;

                if (currentLocation.HasValue)
                    this.DistanceMiles = Location.DistanceMiles(bike.CurrentLocation, currentLocation.Value);
            }
        }
    }
}

File: Web.UI/ClientApp/models/bikes/bikeListItem.ts
import { BikeState } from "./bikeState";
import { BikeModel } from "./bikeModel";
import { Color } from "../master/color";
import { Location } from "../master/location";
import { TypeHelper } from "../../helpers/typeHelper";
import { BikeRef } from "./bikeRef";

export class BikeListItem extends BikeRef {

    public readonly BikeState: BikeState;

    public readonly BikeModel: BikeModel;

    public readonly CurrentLocation: Location;

    public readonly CurrentLocationName: string;

    /// <summary>
    /// Can be available and this is a past date.
    /// Can be unavailable and this is a forecasted future date.
    /// </summary>
    public readonly AvailableFromUtc: string; // Date;

    public readonly CurrentlyAvailable: boolean;

    public readonly RateAverage: number;

    public readonly DistanceMiles: number | null | undefined;
}